pytests/conftest.py

```diff
@@ -0,0 +1,244 @@
+        port = s.getsockname()[1]
+    return port
+
+
+class ErrorHandleMiddleware:
+    def __init__(self, app, pipe=None):
+        self.app = app
+        self.pipe = pipe
+
+    def __call__(self, *args, **kwargs):
+        try:
+            return self.app(*args, **kwargs)
+        except Exception as e:
+            if self.pipe is not None:
+                exc_info = sys.exc_info()
+                exc_type = exc_info[0]
+                tb = exc_info[2]
+                self.pipe.send((exc_type, e, tb))
+            raise e
+
+
+@pytest.hookimpl(hookwrapper=True)
+def pytest_runtest_call(item):
+    yield
+    if client.__name__ in item.funcargs.keys():
+        # check for uncaught errors in server process, raise if any
+        testclient = item.funcargs[client.__name__]
+        testclient._reraise_app_errors()
+
+
+def reraise(f):
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        result = f(*args, **kwargs)
+        testclient = args[0]
+        testclient._reraise_app_errors()
+        return result
+
+    return wrapper
+
+
+class Client(object):
+    host = None
+    port = None
+
+    _error_publisher = None
+    _error_receiver = None
+    _proc = None
+    _session = None
+    _sock = None
+
+    def __init__(self):
+        self.reuse_port = False
+        self.listen_backlog = bjoern.DEFAULT_LISTEN_BACKLOG
+        self._children = []
+
+    @property
+    def session(self):
+        return self._session
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, tb):
+        self.stop()
+
+    def start(self, wsgi_app, num_workers=1):
+        assert self._sock is not None
+        num_workers = max(num_workers, 1)
+        (self._error_receiver, self._error_publisher) = mp.Pipe(False)
+        for _ in range(num_workers):
+            errorhandler = ErrorHandleMiddleware(wsgi_app, self._error_publisher)
+            child = mp.Process(
+                target=bjoern.server_run, args=(self._sock, errorhandler)
+            )
+            child.start()
+            self._children.append(child)
+
+    def stop(self):
+        if self._sock is None:
+            return
+        self.session.close()
+        self._error_publisher.close()
+        self._error_receiver.close()
+        for child in self._children:
+            os.kill(child.pid, signal.SIGTERM)
+            child.join()
+        if self._sock.family == socket.AF_UNIX:
+            filename = self._sock.getsockname()
+            if filename[0] != '\0':
+                os.unlink(self._sock.getsockname())
+        self._sock.close()
+
+    def _reraise_app_errors(self, timeout=0):
+        exc_info = None
+        try:
+            if self._error_receiver.poll(timeout):
+                exc_info = self._error_receiver.recv()
+        except EOFError:
+            return
+        if exc_info is not None:
+            raise_(*exc_info)  # python2 compat
+
+    @reraise
+    def get(self, path=None, **kwargs):
```

Comment:
Maybe refactor these into a single method and use `functools.partial`? But that's nitpicking

Reply:
No, I agree on that, good suggestion. I have mainly added the method copies because of code completion in IDE, but typing stubs would be way better here.